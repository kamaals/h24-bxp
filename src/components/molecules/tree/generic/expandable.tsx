import React from "react";
import * as AccordionPrimitive from "@radix-ui/react-accordion";
import { cn } from "@/lib/utils";
import { ChevronRight } from "lucide-react";

export const ExpandableContent = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Content>,
  React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Content>
>(({ className, children, ...props }, ref) => (
  <AccordionPrimitive.Content
    ref={ref}
    className={cn(
      "overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down",
      className,
    )}
    {...props}
  >
    <div className="pt-0">{children}</div>
  </AccordionPrimitive.Content>
));

ExpandableContent.displayName = AccordionPrimitive.Content.displayName;

export const ExpandableTrigger = React.forwardRef<
  React.ElementRef<typeof AccordionPrimitive.Trigger>,
  React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Trigger>
>(({ className, children, ...props }, ref) => (
  <AccordionPrimitive.Header className="bg-zinc-50 h-10 flex items-center border rounded-l-lg border-r-0">
    <AccordionPrimitive.Trigger
      ref={ref}
      className={cn(
        "accordion-trigger  flex items-center transition-all first:[&[data-state=open]>svg]:rotate-90",
        className,
      )}
      {...props}
    >
      <ChevronRight
        className={cn(
          "h-4 w-4 shrink-0 ml-2 transition-transform duration-200 text-accent-foreground/50 mr-1 relative z-20",
        )}
      />

      {children}
    </AccordionPrimitive.Trigger>
  </AccordionPrimitive.Header>
));

ExpandableTrigger.displayName = AccordionPrimitive.Trigger.displayName;
