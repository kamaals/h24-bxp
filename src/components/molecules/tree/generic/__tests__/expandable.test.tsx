import { describe, it, expect, beforeEach } from "vitest";
import { render, cleanup } from "@testing-library/react";
import { ExpandableContent, ExpandableTrigger } from "../expandable";
import { Accordion, AccordionItem } from "@radix-ui/react-accordion";

describe(ExpandableContent.name, () => {
  beforeEach(() => {
    cleanup();
  });

  it("Should render correctly for close", () => {
    const { container } = render(
      <Accordion type="single" collapsible>
        <AccordionItem value={"item-1"}>
          <ExpandableContent>
            <div data-testid="action">Action</div>
          </ExpandableContent>
        </AccordionItem>
      </Accordion>,
    );
    expect(
      container.querySelector('div[data-state="closed"]'),
    ).toBeInTheDocument();
  });

  it("Should render correctly for open", () => {
    const { container } = render(
      <Accordion type="single" collapsible>
        <AccordionItem value={"item-1"} data-state="open">
          <ExpandableTrigger>Open</ExpandableTrigger>
          <ExpandableContent>
            <div data-testid="action">Action</div>
          </ExpandableContent>
        </AccordionItem>
      </Accordion>,
    );
    expect(
      container.querySelector('div[data-state="open"]'),
    ).toBeInTheDocument();
  });

  it("Should match the snapshot", () => {
    const resp = render(
      <Accordion type="single" collapsible>
        <AccordionItem value={"item-1"}>
          <ExpandableTrigger>Open</ExpandableTrigger>
          <ExpandableContent>
            <div data-testid="action">Action</div>
          </ExpandableContent>
        </AccordionItem>
      </Accordion>,
    );
    expect(resp).toMatchSnapshot();
  });
});
