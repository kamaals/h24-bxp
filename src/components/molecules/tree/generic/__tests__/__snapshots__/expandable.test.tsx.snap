// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`undefined > Should match the snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-orientation="vertical"
      >
        <div
          data-orientation="vertical"
          data-state="closed"
        >
          <h3
            class="bg-zinc-50 h-10 flex items-center border rounded-l-lg border-r-0"
            data-orientation="vertical"
            data-state="closed"
          >
            <button
              aria-controls="radix-«r5»"
              aria-expanded="false"
              class="accordion-trigger flex items-center transition-all first:[&[data-state=open]>svg]:rotate-90"
              data-orientation="vertical"
              data-radix-collection-item=""
              data-state="closed"
              id="radix-«r4»"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="lucide lucide-chevron-right h-4 w-4 shrink-0 ml-2 transition-transform duration-200 text-accent-foreground/50 mr-1 relative z-20"
                fill="none"
                height="24"
                stroke="currentColor"
                stroke-linecap="round"
                stroke-linejoin="round"
                stroke-width="2"
                viewBox="0 0 24 24"
                width="24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="m9 18 6-6-6-6"
                />
              </svg>
              Open
            </button>
          </h3>
          <div
            aria-labelledby="radix-«r4»"
            class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
            data-orientation="vertical"
            data-state="closed"
            hidden=""
            id="radix-«r5»"
            role="region"
            style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-orientation="vertical"
    >
      <div
        data-orientation="vertical"
        data-state="closed"
      >
        <h3
          class="bg-zinc-50 h-10 flex items-center border rounded-l-lg border-r-0"
          data-orientation="vertical"
          data-state="closed"
        >
          <button
            aria-controls="radix-«r5»"
            aria-expanded="false"
            class="accordion-trigger flex items-center transition-all first:[&[data-state=open]>svg]:rotate-90"
            data-orientation="vertical"
            data-radix-collection-item=""
            data-state="closed"
            id="radix-«r4»"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="lucide lucide-chevron-right h-4 w-4 shrink-0 ml-2 transition-transform duration-200 text-accent-foreground/50 mr-1 relative z-20"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="m9 18 6-6-6-6"
              />
            </svg>
            Open
          </button>
        </h3>
        <div
          aria-labelledby="radix-«r4»"
          class="overflow-hidden text-sm transition-all data-[state=closed]:animate-accordion-up data-[state=open]:animate-accordion-down"
          data-orientation="vertical"
          data-state="closed"
          hidden=""
          id="radix-«r5»"
          role="region"
          style="--radix-accordion-content-height: var(--radix-collapsible-content-height); --radix-accordion-content-width: var(--radix-collapsible-content-width);"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
